File: heapsort/javascript/maxheap.js
export function fastHeapify(a) {
    const end = a.length - 1;
    for (let i = parent(end); i >= 0; i -= 1) {
        bubbleDown(a, i, end);
    }
    return new Heap(a, end);
}

export function slowHeapify(a) {
    const end = a.length - 1;
    for (let i = 1; i <= end; i += 1) {
        bubbleUp(a, i, end);
    }
    return new Heap(a, end);
}

class Heap {
    constructor(a, end) {
        this.a = a;
        this.end = end;
    }

    deleteMax() {
        const maxValue = this.a[0];
        this.a[0] = this.a[this.end];
        this.end -= 1;
        bubbleDown(this.a, 0, this.end);
        return maxValue;
    }
}

function bubbleUp(a, i) {
    const p = parent(i);
    if (p >= 0 && a[p] < a[i]) {
        swap(a, p, i);
        bubbleUp(a, p);
    }
}

function bubbleDown(a, i, end) {
    let max = i;

    const l = leftChild(i);
    if (l <= end && a[l] > a[max]) {
        max = l;
    }

    const r = rightChild(i);
    if (r <= end && a[r] > a[max]) {
        max = r;
    }

    if (max !== i) {
        swap(a, max, i);
        bubbleDown(a, max, end);
    }
}

function swap(a, i, j) {
    const t = a[i];
    a[i] = a[j];
    a[j] = t;
}

function parent(i) {
    return Math.floor((i - 1) / 2);
}

function leftChild(i) {
    return 2 * i + 1;
}

function rightChild(i) {
    return 2 * i + 2;
}

File: heapsort/go/maxheap.go
package heapsort

type MaxHeap []int
type node int

type MaxHeapifyFunc func(a []int) MaxHeap

func SlowMaxHeapify(a []int) MaxHeap {
	h := (MaxHeap)(a)
	for i := range h {
		h.bubbleUp(node(i))
	}
	return h
}

func FastMaxHeapify(a []int) MaxHeap {
	h := (MaxHeap)(a)
	for i := parent(h.last()); i >= 0; i-- {
		h.bubbleDown(node(i))
	}
	return h
}

func (h *MaxHeap) DeleteMax() int {
	a := (*h)
	maxValue := a[0]

	l := a.last()
	a[0] = a[l]
	(*h) = (*h)[:l]
	h.bubbleDown(0)

	return maxValue
}

func (h MaxHeap) last() node {
	return node(len(h) - 1)
}

func (h MaxHeap) bubbleUp(n node) {
	if n == 0 {
		return
	}
	if p := parent(n); h[p] < h[n] {
		h.swap(n, p)
		h.bubbleUp(p)
	}
}

func (h MaxHeap) bubbleDown(n node) {
	max := n
	if l := leftChild(n); l <= h.last() && h[l] > h[max] {
		max = l
	}
	if r := rightChild(n); r <= h.last() && h[r] > h[max] {
		max = r
	}

	if max != n {
		h.swap(max, n)
		h.bubbleDown(max)
	}
}

func (h MaxHeap) swap(i, j node) {
	h[i], h[j] = h[j], h[i]
}

func parent(n node) node {
	return (n - 1) / 2
}

func leftChild(n node) node {
	return 2*n + 1
}

func rightChild(n node) node {
	return 2*n + 2
}
